jwt/jwt.go

```diff
@@ -0,0 +1,128 @@
+package jwt
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+)
+
+type header struct {
+	// The algorithm used for signature.
+	Algorithm string `json:"alg"`
+	// Represents the token type.
+	Typ string `json:"typ"`
+	// The hint which key is being used.  ID Tokens SHOULD NOT use the JWS or
+	// JWE x5u, x5c, jku, or jwk Header Parameter fields. Instead, references to
+	// keys used are communicated in advance using Discovery and Registration
+	// parameters, per Section 10.
+	KeyID string `json:"kid"`
+}
+
+type ClaimSet struct {
+	Issuer    string   `json:"iss,omitempty"`
+	Subject   string   `json:"sub,omitempty"`
+	Audience  []string `json:"aud,omitempty"`
+	Expiry    int64    `json:"exp,omitempty"`
+	NotBefore int64    `json:"nbf,omitempty"`
+	IssuedAt  int64    `json:"iat,omitempty"`
+	JwtId     string   `json:"jti,omitempty"`
+}
+
+func EncodeAndSign(claims interface{}, keyID string, privateKey *ecdsa.PrivateKey) (string, error) {
+	header := header{
+		Algorithm: "ES256",
+		Typ:       "JWT",
+		KeyID:     keyID,
+	}
+	headerJSON, err := json.Marshal(header)
```

Comment:
Passing a `*header` here simplifies the conversion to the `interface{}`

Reply:
I'm not sure I follow. Are you saying I should just drop the reference type and use a value type instead?

Reply:
No, I mean:
```
	header := &header{
		Algorithm: "ES256",
		Typ:       "JWT",
		KeyID:     keyID,
	}
	headerJSON, err := json.Marshal(header)
```